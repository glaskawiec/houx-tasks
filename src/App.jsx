import React from 'react';
import { useHoux } from 'houx';
import { Grid } from 'react-flexbox-grid';
import AddOrEditTaskModal from './components/AddOrEditTaskModal';
import Header from './components/Header';
import TasksTable from './components/TasksTable';
import 'bootstrap/dist/css/bootstrap.min.css';
import GlobalStyle from './GlobalStyle';
import AddTaskButton from './components/AddTaskButton';

const App = () => {
  const { state } = useHoux();

  return (
    <Grid fluid>
      <GlobalStyle />
      <Header />
      {state.tasks.isAddModalOpen ? <AddOrEditTaskModal /> : null}
      {state.tasks.isEditModalOpen
        ? <AddOrEditTaskModal editTaskId={state.tasks.currentEditTaskId} />
        : null}
      <TasksTable />
      <AddTaskButton />
    </Grid>
  );
};

export default App;
