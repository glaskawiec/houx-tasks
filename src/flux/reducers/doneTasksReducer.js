import { DONE_TASKS_ADD, DONE_TASKS_REMOVE } from "../actionTypes/doneTasksActionTypes";

export const initialState = [];

const doneTasksReducer = (state = initialState, action) => {
    switch (action && action.type) {
        case DONE_TASKS_ADD:
            return [...state, action.task]
        case DONE_TASKS_REMOVE:
            return state.filter((task) => task.id !== action.id)
        default: return state;
    }
}

export default doneTasksReducer;