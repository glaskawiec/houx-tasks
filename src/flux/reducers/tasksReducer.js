import {
  TASKS_ADD,
  TASKS_REMOVE,
  TASKS_COMPLETE,
  TASKS_EDIT,
  TASKS_OPEN_ADD_MODAL,
  TASKS_OPEN_EDIT_MODAL,
  TASKS_CLOSE_MODAL, TASKS_CHANGE_ORDER,
} from '../actionTypes/tasksActionTypes';

export const initialState = {
  tasks: [
    {
      id: 0,
      title: 'Example task',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      colors: {
        text: 'pink',
      },
    },
    {
      id: 1,
      title: 'Example task 2',
      description: '',
      colors: {
        text: 'purple',
      },
    },
    {
      id: 2,
      title: 'Example task 3',
      description: '',
      colors: {
        text: 'yellow',
      },
    },
  ],
  nextId: 3,
  isEditModalOpen: false,
  currentEditTaskId: null,
  isAddModalOpen: false,
};

const tasksReducer = (state = initialState, action) => {
  switch (action && action.type) {
    case TASKS_ADD: {
      const newTodo = {
        ...action.task,
        id: state.nextId,
      };

      return {
        ...state,
        nextId: state.nextId + 1,
        tasks: [...state.tasks, newTodo],
      };
    }
    case TASKS_REMOVE: {
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    }
    case TASKS_COMPLETE:
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    case TASKS_OPEN_ADD_MODAL:
      return {
        ...state,
        isAddModalOpen: true,

      };
    case TASKS_CLOSE_MODAL:
      return {
        ...state,
        isEditModalOpen: false,
        isAddModalOpen: false,
      };
    case TASKS_OPEN_EDIT_MODAL:
      return {
        ...state,
        isEditModalOpen: true,
        currentEditTaskId: action.id,
      };
    case TASKS_EDIT: {
      const { currentEditTaskId } = state;
      return {
        ...state,
        tasks: [
          ...state.tasks.filter((task) => task.id !== action.id),
          { ...action.editedTask, id: currentEditTaskId },
        ],
        currentEditTaskId: null,
      };
    }
    case TASKS_CHANGE_ORDER: {
      const {orderDown, id} = action;
      const tasksCopy = [...state.tasks];
      let tasksToOrderIndex = 0;

      for( ; tasksToOrderIndex < tasksCopy.length; ++tasksToOrderIndex){
        if(tasksCopy[tasksToOrderIndex].id === id){
          break;
        }
      }

      if(orderDown){
        const previousId = tasksCopy[tasksToOrderIndex+1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex+1].id = id;
      } else {
        const previousId = tasksCopy[tasksToOrderIndex-1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex-1].id = id;
      }

      return {
        ...state,
        tasks: tasksCopy
      };
    }
    default: return state;
  }
};

export default tasksReducer;
