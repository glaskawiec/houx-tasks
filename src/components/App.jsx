import React from 'react';
import TasksWrapper from './Routes/Tasks/TasksWrapper';
import reducers from '../flux/reducers';
import { Provider, createStoreWithReducers } from 'houx';
import Layout from './Layout/Layout';
import { BrowserRouter, Route } from 'react-router-dom'
import DoneTasksTable from './Routes/DoneTasks/DoneTasksTable';

export default function App() {
  const store = createStoreWithReducers(reducers);
  return (
    <Provider store={store}>
      <BrowserRouter>
        <Layout>
          <Route exact path="/" component={TasksWrapper} />
          <Route path="/done" component={DoneTasksTable} />
        </Layout>
      </BrowserRouter>
    </Provider>
  )
}
