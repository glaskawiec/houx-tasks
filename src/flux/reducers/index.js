import tasksReducer from './tasksReducer';

const reducers = {
  tasks: tasksReducer,
};

export default reducers;
