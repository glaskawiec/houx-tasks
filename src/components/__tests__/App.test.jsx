import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { HouxProvider } from 'houx';
import reducers from '../../flux/reducers';
import App from '../../App';
import tasksReducer from '../../flux/reducers/tasksReducer';

describe('App component unit tests', () => {
  const mockedState = {
    tasks: [
      {
        id: 0,
        title: 'Test task 1',
        description: 'Test description 1',
        colors: {
          text: 'pink',
          background: 'yellow',
        },
      },
    ],
    nextId: 1,
    isEditModalOpen: false,
    currentEditTaskId: null,
    isAddModalOpen: false,
  };

  const mockedReducers = {
    tasks: (state, action) => tasksReducer(mockedState, action),
  };

  test('App should render properly', () => {
    const { container } = render(
      <HouxProvider reducers={reducers}>
        <App />
      </HouxProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Clicking on add task button should show add task modal with empty inputs', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={reducers}>
        <App />
      </HouxProvider>,
    );

    fireEvent.click(getByTestId('add-task-button'));

    expect(getByTestId('add-or-edit-task-modal')).toBeInTheDocument();

    expect(getByTestId('title-input').value).toBe('');
    expect(getByTestId('description-input').value).toBe('');
    expect(getByTestId('text-color-input').value).toBe('');
    expect(getByTestId('background-color-input').value).toBe('');
  });

  test('Clicking on edit task button should show edit modal with proper data', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <App />
      </HouxProvider>,
    );

    fireEvent.click(getByTestId('edit-task-0-button'));

    expect(getByTestId('add-or-edit-task-modal')).toBeInTheDocument();

    expect(getByTestId('title-input').value).toBe(mockedState.tasks[0].title);
    expect(getByTestId('description-input').value).toBe(mockedState.tasks[0].description);
    expect(getByTestId('text-color-input').value).toBe(mockedState.tasks[0].colors.text);
    expect(getByTestId('background-color-input').value).toBe(mockedState.tasks[0].colors.background);
  });
});
