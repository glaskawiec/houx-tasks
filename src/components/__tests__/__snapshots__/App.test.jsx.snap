// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component unit tests App should render properly 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <header>
      <h2
        class="sc-AxgMl ievzTO"
      >
        Task Manager
      </h2>
    </header>
    <div
      class="table-responsive"
    >
      <table
        class="table table-dark table-bordered"
      >
        <thead>
          <tr>
            <th>
              Id
            </th>
            <th>
              Title
            </th>
            <th>
              Description
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="sc-fzozJi ekjuyZ"
            color="pink"
            data-testid="table-row-0"
          >
            <td
              data-testid="table-id-cell-0"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                0
              </p>
            </td>
            <td
              data-testid="table-title-cell-0"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                Example task
              </p>
            </td>
            <td
              class="sc-AxmLO dweuuQ"
              data-testid="table-description-cell-0"
            >
              <span
                class="sc-AxheI GGwlK"
                data-testid="collapse-button-task-0"
              >
                ➡
              </span>
            </td>
            <td>
              <span
                aria-label="Cross"
                aria-labelledby="Cross"
                data-testid="remove-task-0-button"
                role="img"
                style="cursor: pointer;"
              >
                ❌
              </span>
              <span
                aria-label="Cross"
                data-testid="edit-task-0-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ✏
              </span>
              <span
                aria-label="arrow up"
                data-testid="order-down-task-0-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ⬇
                ️
              </span>
            </td>
          </tr>
          <tr
            class="sc-fzozJi bLtv"
            color="purple"
            data-testid="table-row-1"
          >
            <td
              data-testid="table-id-cell-1"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                1
              </p>
            </td>
            <td
              data-testid="table-title-cell-1"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                Example task 2
              </p>
            </td>
            <td
              class="sc-AxmLO dweuuQ"
              data-testid="table-description-cell-1"
            />
            <td>
              <span
                aria-label="Cross"
                aria-labelledby="Cross"
                data-testid="remove-task-1-button"
                role="img"
                style="cursor: pointer;"
              >
                ❌
              </span>
              <span
                aria-label="Cross"
                data-testid="edit-task-1-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ✏
              </span>
              <span
                aria-label="arrow up"
                data-testid="order-down-task-1-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ⬇
                ️
              </span>
              <span
                aria-label="arrow down"
                data-testid="order-up-task-1-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ⬆
                ️
              </span>
            </td>
          </tr>
          <tr
            class="sc-fzozJi ctflPh"
            color="yellow"
            data-testid="table-row-2"
          >
            <td
              data-testid="table-id-cell-2"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                2
              </p>
            </td>
            <td
              data-testid="table-title-cell-2"
            >
              <p
                class="sc-Axmtr hFrjGC"
              >
                Example task 3
              </p>
            </td>
            <td
              class="sc-AxmLO dweuuQ"
              data-testid="table-description-cell-2"
            />
            <td>
              <span
                aria-label="Cross"
                aria-labelledby="Cross"
                data-testid="remove-task-2-button"
                role="img"
                style="cursor: pointer;"
              >
                ❌
              </span>
              <span
                aria-label="Cross"
                data-testid="edit-task-2-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ✏
              </span>
              <span
                aria-label="arrow down"
                data-testid="order-up-task-2-button"
                role="img"
                style="margin-left: 0.5em; cursor: pointer;"
              >
                ⬆
                ️
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="btn btn-primary"
      data-testid="add-task-button"
      type="button"
    >
      Add Task
    </button>
  </div>
</div>
`;
